src/Elastic.CommonSchema.NLog/EcsLayout.cs

```diff
@@ -0,0 +1,400 @@
+﻿// Licensed to Elasticsearch B.V under one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using NLog;
+using NLog.Config;
+using NLog.Layouts;
+using NLog.Targets;
+
+namespace Elastic.CommonSchema.NLog
+{
+	public class EcsLayout : Layout
+	{
+		public EcsLayout()
+		{
+			IncludeAllProperties = true;
+
+			ProcessId = "${processid}";
+			ProcessName = "${processname:FullName=false}";
+			ProcessExecutable = "${processname:FullName=true}";
+			ProcessTitle = "${processinfo:MainWindowTitle}";
+			ProcessThreadId = "${threadid}";
+
+			HostName = "${machinename}";
+			HostIp = "${local-ip:cachedSeconds=60}"; // NLog 4.6.8
+
+			ServerUser = "${environment-user}"; // NLog 4.6.4
+
+			// These values are set by the Elastic.Apm.NLog package
+			ApmTraceId = "${ElasticApmTraceId}";
+			ApmTransactionId = "${ElasticApmTransactionId}";
+		}
+
+		private static class SpecialKeys
+		{
+			public const string DefaultLogger = "Elastic.CommonSchema.NLog";
+		}
+
+		public Layout AgentId { get; set; }
+		public Layout AgentName { get; set; }
+		public Layout AgentType { get; set; }
+		public Layout AgentVersion { get; set; }
+		public Layout ApmTraceId { get; set; }
+		public Layout ApmTransactionId { get; set; }
+
+		public Layout EventAction { get; set; }
+		public Layout EventCategory { get; set; }
+		public Layout EventId { get; set; }
+		public Layout EventKind { get; set; }
+		public Layout EventSeverity { get; set; }
+
+		public Layout HostId { get; set; }
+		public Layout HostIp { get; set; }
+		public Layout HostName { get; set; }
+
+		public bool IncludeAllProperties { get; set; }
+
+		[ArrayParameter(typeof(TargetPropertyWithContext), "label")]
+		public IList<TargetPropertyWithContext> Labels { get; private set; } = new List<TargetPropertyWithContext>();
+
+		[ArrayParameter(typeof(TargetPropertyWithContext), "metadata")]
+		public IList<TargetPropertyWithContext> MetaData { get; private set; } = new List<TargetPropertyWithContext>();
+
+		public Layout ProcessExecutable { get; set; }
+		public Layout ProcessId { get; set; }
+		public Layout ProcessName { get; set; }
+		public Layout ProcessThreadId { get; set; }
+		public Layout ProcessTitle { get; set; }
+
+		public Layout ServerAddress { get; set; }
+		public Layout ServerIp { get; set; }
+		public Layout ServerUser { get; set; }
+
+		[ArrayParameter(typeof(TargetPropertyWithContext), "tag")]
+		public IList<TargetPropertyWithContext> Tags { get; private set; } = new List<TargetPropertyWithContext>();
+
+		protected override void RenderFormattedMessage(LogEventInfo logEventInfo, StringBuilder target)
+		{
+			var exceptions = logEventInfo.Exception != null
```

Comment:
Why the need to allocate the `List<Exception>` ? There is no logic that requires a List. `GetError` just takes first item. Could one just pass the `logEventInfo.Exception` as parameter to `GetError` ? (Same pattern also exists in the Serilog, and there it also doesn't make sense).

Reply:
Looks like an unnecessary allocation, will remove.